### ISSUE TITLE
Cooperation consultation and some serious problem

### ISSUE THREAD

**Issue by fangjialong2019** [ONLYOFFICE/DocumentServer#870]
We are using your community version of OnlyOffice Document Server,
We have some problems, as follows:
1. It is impossible to undo and redo the multi-player collaboration in the fast mode
2, the skin is not very beautiful, not friendly to the Chinese，It is best to be able to provide a UI similar to WPS or MS Office
3, Use the mouse to move the scroll bar, can make the table move smoothly? Not move by row step.
If you can help us solve the above problems, we may consider buying the paid version.
If the above request can be solved, please tell us the price.

The same content was sent to sales@onlyoffice.com

**Comment by fangjialong2019** [comment]
Hope to reply as soon as possible.

**Comment by Rita-Bubnova** [comment]
Hello, @fangjialong2019.
Thank you for report.
1. I can confirm - this is a enhancement, issues 45357, 45358, 45359, 45360 in our internal issue tracker.
2. Would you report in more detail what it means "the skin is not very beautiful, not friendly to the Chinese"?
3. To this question I answered you here https://github.com/ONLYOFFICE/DocumentServer/issues/736#issuecomment-627229111

**Comment by fangjialong2019** [comment]
@Rita-Bubnova 
Excuse me, What does ETA mean？

To the second question，the people around us are used to using Google Doc and MS Office, so we want OnlyOffice to have a similar experience.

To the third question，You asked me to contact sales to speed up it.  
it means we can paid for it to spped up?

**Comment by fangjialong2019** [comment]
To the Undo-Redo question , This is my solution as follows.
The following upgrade files enable Excel‘s undo-redo functionality.
Usage:
Place the file in the sdkjs.cell directory and compile it with the api files.And then modify the webapp. Spreadsheeteditor. Main. Js file, and start it in the function of onDocumentContentReady

Results:
Most of cases have been supported Undo-Redo in fast co-edit mode.But in rare cases, data can be lost and we can't reproduce it

[cell-api-upgrade.zip](https://github.com/ONLYOFFICE/DocumentServer/files/4616123/cell-api-upgrade.zip)

**Comment by Rita-Bubnova** [comment]
@fangjialong2019.
ETA this is the abbreviation for Estimated time of arrival, that is, we cannot tell the deadline for the implementation this enhancement.

> To the second question，the people around us are used to using Google Doc and MS Office, so we want OnlyOffice to have a similar experience.

This is very general statement, I don't know that you really mean by this. And we can try to add feature you're missing, but we are separate product from MS Office and never be the same.

> To the third question，You asked me to contact sales to speed up it.
it means we can paid for it to spped up?

I think so, you should discuss any feature you want to pay with sales team, money questions is out of my responsibilities.

**Comment by fangjialong2019** [comment]
To the Undo-Redo question for word and slide.
Usage is similar to cell.

[word-slide-upgrade.zip](https://github.com/ONLYOFFICE/DocumentServer/files/4616216/word-slide-upgrade.zip)

**Comment by fangjialong2019** [comment]
@Rita-Bubnova 

UI is not a big problem, we can add skins based on source code of the community version.
Thank you for your contribution to open source.

**Comment by fangjialong2019** [comment]
@Rita-Bubnova 

UI is not a big problem, we can add skins based on source code of the community version.
Thank you for your contribution to open source.

If possible, please help me to check my solution to the question of Undo-Redo

**Comment by ShockwaveNN** [comment]
Ouur dev team is not ready to implement undo-redo in fast co-edit
As you mention in your case data loss can happen and this code cannot be released in production-ready solution as ours

And since our dev team is not ready - I don't think any one in our team can provide you help with your solution

**Comment by fangjialong2019** [comment]
@ShockwaveNN 
I hope you can solve this problem quickly, and you can refer to my solution, because it works fine most of the time.

**Comment by RuslanGabbasov** [comment]
@ShockwaveNN In strict collaborative mode, that we should to use because otherwise collaborative usage of spreadsheets is dangerous, data loses very often too. In the reason of network connection lost or computer goes into sleep mode when user write something without manual saving. Most of our clients denied of use spreadsheet editor in OnlyOffice only because of this.
I think you should increase priority of this feature to absolutely immediately important level.

**Comment by ShockwaveNN** [comment]
@RuslanGabbasov Sorry to hear this

As far as I understand it - implementing undo-redo in co-edit mode is pretty complicated task, so I think it's main reason why it's still not implemented

A shown your report to our dev team, but I'm not in position to change priorities of feature to implement

**Comment by fangjialong2019** [comment]
@RuslanGabbasov For this reason, we purchased WPS WebOffice

**Comment by biva** [comment]
Dear @ShockwaveNN and @Rita-Bubnova 
Thank you for this great OnlyOffice: our team is really impressed by the quality, the reactivity and the look of your solution :) We love it.
The undo-redo feature in fast co-editing mode would really be great indeed. I understand it's not an easy task, but do you have some good news in this regard? Thank you again !

**Comment by ShockwaveNN** [comment]
@biva Sorry, no news

Since we are in no control of priority of this task - if this is very important for you - you can try to contact sales team at sales@onlyoffice.com and discuss it

**Comment by biva** [comment]
@ShockwaveNN thanks for your quick feedback: it's not that urgent, it's just that I pressed Ctrl+Z several times and was surprised to not see this working ;)
I understand: you have to deal with priorities. I hope it will come soon :) And congrats again for your great product, it's even better than Google Sheet!

**Comment by ImaCrea** [comment]
Hello dear people at OnlyOffice.

another message here to tell you this cmd+z feature is much needed here also. Especially on spreadsheet where people work on formulas and complex data.

Users feedbacks we collected go along these lines:
- the modal shown is scary, people think they're gonna lose some data. it's hard to understand and "strict mode" wording isn't very helpful neither.
- once in strict mode, many people don't understand they need to click on the save icon to actually push new data to other connected users.
- even less people understand they need to go into settings to switch back to "fast mode" if they're tired of clicking on save button every time they do something. also "fast mode" wording is not very clear and people don't really get what it means.

From what I understand, there's a technical challenge behind that which seems quiet big. That might explain this complicated experience and why you haven't been able to prioritize that issue for the moment.

I understand technically it's complicated at the moment to prioritize and this is not in your backlog (or maybe it changed since latest message here?), but maybe we could find an intermediate solution by focusing on ux side first?

I'm curious if you thought about such angle to try to solve that problem?

**Comment by ShockwaveNN** [comment]
@ImaCrea Thanks for your details explanation, I've pointed out it to our UX team, maybe they have some ideas

**Comment by biva** [comment]
Hello @ShockwaveNN , would you have any good news on this issue?

**Comment by ShockwaveNN** [comment]
Sorry no news

**Comment by ImaCrea** [comment]
> Sorry no news

Still no news by any chance? ^^'

**Comment by ShockwaveNN** [comment]
@ImaCrea Sorry, no news. This is a very-very complicated task and currently not our top priority. If you want to discuss our priorities - feel free to contact sales@onlyoffice.com

**Comment by ImaCrea** [comment]
> @ImaCrea Sorry, no news. This is a very-very complicated task and currently not our top priority. If you want to discuss our priorities - feel free to contact [sales@onlyoffice.com](mailto:sales@onlyoffice.com)

Alright, I'll think about it. Do you have a public roadmap somewhere by any chance please?

**Comment by ShockwaveNN** [comment]
@ImaCrea Yes, [we have](https://github.com/ONLYOFFICE/DocumentServer/blob/master/ROADMAP.md)
It's not always most actual one, but it's the best we can provide

**Comment by ImaCrea** [comment]
Hello there,

We're on 6.4.2 and I can confirm Undo-Redo is now working on fast co-edit mode for docx and pptx. This is great news.

Would be amazing when it work for xlsx!

**Comment by yurkobb** [comment]
I see that in the desktop version of the spreadsheet editor version 7.1.0.215 Undo & Redo works in Fast co-editing mode (that is great news!). Is it coming to the web version as well?

**Comment by ShockwaveNN** [comment]
@yurkobb Yes, this issue is still unresolved fully

**Comment by PasserDreamer** [comment]
I don't know how OnlyOffice implemented this function.
How about follow what git do? IMO, every change is a commit, everyone who is opening the file must sync to head before save changes (branch migration). Undo and Redo also as a new commit.
